File: frontend/src/types/Messages.ts
import type { Game } from "./Game";
import type { Player } from "./Player";
import type { Tasks } from "./Tasks";

export enum IncomingMessageType {
  NewTask = "newTask",
  PlayerUpdated = "playerUpdated",
  GameUpdated = "gameUpdated",
}

export type IncomingMessages = {
  [IncomingMessageType.NewTask]: NewTask;
  [IncomingMessageType.PlayerUpdated]: PlayerUpdated;
  [IncomingMessageType.GameUpdated]: GameUpdated;
};

export type NewTask = { task: Tasks };
export type PlayerUpdated = { player: Player };
export type GameUpdated = { game: Game };

File: backend/src/model/messages.rs
use super::{Game, Player, TaskDefinition};
use serde::{Deserialize, Serialize};

/// All incoming message types which might be send by peers and should be handled on the server side.
#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum IncomingMessage {
    Auth { token: String },
}

/// All outgoing message types which might be send to the peers.
#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum OutgoingMessage {
    Welcome {},
    NewTask { task: TaskDefinition },
    PlayerUpdated { player: Player },
    PlayerLeft { player_id: String },
    PlayerAttended { player: Player },
    GameUpdated { game: Game },
}
